Update time-remaining.js: add Playback Rate

muescha opened the pull request:
**What:**
- Added `playbackRate` to the time remaining calculation.

**Questions:**
1) Should this be implemented in the current plugin, or should it be a separate plugin independent of the `time-remaining` plugin? A separate plugin would be code duplication, but might more clear in the options dialog.
2) Should it be added as an additional option to `time-remaing`, similar to "time_remaining_position," with a setting like `show_playbackRate: yes/no`? However, this would limit the customization of its position in the template string.
3) Should `playbackRate` be included in all format templates where speed is relevant?

muescha committed:
Update time-remaining.jsad

muescha commented:
![Bildschirmfoto 2024-08-09 um 14 46 06](https://github.com/user-attachments/assets/fc584b1b-92c7-4235-98bf-29fa82d86d4b)

![Bildschirmfoto 2024-08-09 um 14 47 01](https://github.com/user-attachments/assets/2efa91cb-f879-4a36-9e07-7fe65ca6f88d)

raingart commented:
Hi @muescha 

the current implementation involves choosing from a given list of templates.
But there is an alternative option. Which was not included for reasons of obviousness to users. In it the user can enter the selected format himself

https://github.com/raingart/Nova-YouTube-extension/blob/master/plugins/control-panel/time-remaining.js#L139-L179

You can try uncommenting. If this option suits you, then I will leave it. I assumed that the user is stupid and will not be able to figure it out. It’s easier and safer to choose from a ready-list. Because there are no visual hints for the user. And as you previously issue, there are quite big problems with the Wiki. And it was difficult for me to convey the possibilities clearly and clearly to the user. And laziness of course.

If you have any problems or questions, write here or by email, I will try to answer and help you.

muescha commented:
yes - the first strategy with user edited templates are a lot more complicated.

- the input field is to small (in my case):

![Bildschirmfoto 2024-08-09 um 19 30 53](https://github.com/user-attachments/assets/7ffae465-4f79-47fc-85d3-b3fa576c6d9c)

- the dropdown is not working in my case - I don't invested too much time to investigage. but a short hint I found in the console:
```
GET https://raingart.github.io/nova/js/options.js net::ERR_ABORTED 404 (Not Found)
```
I opened a new issue: #185

- I would also do a more template aproach - put each value in braches `{}` - but also with this the template is not easy to read:

`{left}/{left*speed} ({done%}) {speed}x`

muescha commented:
I think the current state is better to understand by an user.

so leaving the question 1-3: 
- a separate plugin would be more easy to understand, because it is an on/off option
- adding it to the template string results that we need to add it also to more of the templates which results in a long template list
- adding it as an option would result that the playback rate is only visible when time remaining is switched on

it is up to you what would be the best implementation.

muescha commented:
I generated the userscript right from the github repo (created a script: #186) 
what is different to the dev ver?

muescha commented:
with the video I got it: it shows nothing because the dropdown was filtered by the exisiting template, because of no match there was no value.

I would redesign the template filed in the form of:
- textfield (or textarea) which is more wide (maybe start on a new line)
- the "texteditor" for this textarea should have 2 buttons:
	- `Use Templates`: there is a dropdown with some predefined templates which replaces the template string (this is your strategy 2)
	- `Insert field`: there is a dropdown of all available fields (the dropdown should not filtered) which are inserted at the cursor:
	
| Label                                    | Value                                    | Example |
|------------------------------------------|--------------------------------------|-|
| speed 	| {speed} | 1.5 |
| left                                     | {left}                                     | 00:02:30
| left%                                    | {left%}                                    | 25%
| left\*speed                              | {left*speed}                              | 00:01:40
| done                                    | {done}                                    |  00:07:30
| done%                                    | {done%}                                    | 75%
| duration      | {duration}      | 00:10:00
| duration*speed      | {duration*speed}      | 00:06:40

muescha commented:
thx for the `time-remaining.txt` - is the current state in any branch?

raingart commented:
>what is different to the dev ver?

the structure itself remains the same, just a little refactoring and fixing. Since I wrote it myself, `git push` occurred when cumulative changes denoting a new release.

branch now updated

Changelog.md:
- add `Apply if URL has link to comment` option for [video-autopause]
- add `Apply if URL has link to comment` option for [video-autostop]
- add `Custom colors` option for [sponsor-block]
- add `Get screenshot size from` option for [player-quick-buttons]
- merge `clipboard` and `formats` in Screenshot out [player-quick-buttons]
- add `popup` in Screenshot out [player-quick-buttons]
- add `Min iframe width` option for [embed-popup]
- add `speed` option for [time-remaining] by @muescha
- cut from (thumbs-title-normalize) and move `Show full title` in new plugin for [thumbs-title-show-full]
- add `` option for []
- remove `` in []

new plugin
- []

removed plugin:
- []

fixes in plugins:
- [time-remaining]
- [channel-play-all]
- [theater-mode]
- [thumbs-hide]
- [player-quick-buttons] - aspect-ratio, rotate
- [search-filter]
- [video-zoom]
- [rss-link]
- []

optimized plugins:
- [sidebar-thumbs-channel-link-patch]
- [thumbs-clear]
- [thumbs-watch-later]
- [thumbs-not-interested]
- [thumbs-title-filter]
- [video-rate] - recalc fn
- [player-control-autohide]
- [comments-sort]
- []

minor optimized plugins:
- [video-date-format]
- [sponsor-block]
- [player-float-progress-bar]
- [video-rate]
- [save-channel-state]
- [playlist-reverse]
- [video-unblock-region]
- [embed-popup]
- [square-avatars]
- [download-video] - update sources list
- []

raingart commented:
I don't know git very well 
That's why I added you to Collaborators
I think it will be easier for you this way than when I guess to move this commit to a new branch
`git cherry-pick`

and even more so solve “branch has conflicts”!

raingart commented:
change any files and documentation as you see fit

raingart commented:
I simplifying the format a little.
`datalist` is not suitable for displaying data from 3 columns. It is more adapted for 2nd.

I also made a few examples at the beginning (I think they are the most popular, maybe) and then a list of all available commands

![Screenshot from 2024-08-10 15-41-30](https://github.com/user-attachments/assets/3b20959e-c687-4d57-8e1f-c9f17b5bd691)

muescha commented:
That looks great :)

- I created some refactoring - no function change (PR #189)

muescha commented:
I tried also a switch function - and found some problems in `options.js` - i tried to use `'data-dependent': { 'time_remaining_format_use_custom': false },` but this was only possible without the changes in `options.js`

see https://github.com/muescha/Nova-YouTube-extension/commit/542aa105f51df458e8b1f968824dd748298bdc13


### Changes

- checkbox value & string value
```diff
- // return [subtargetEl.value];
+ return [(subtargetEl.type == 'checkbox') ? subtargetEl.checked.toString() : subtargetEl.value];
```
I changed it to the previous comment, because the checkbox always have the value `on` regardless if is checked or unchecked. I added a `.toString()` because the comparison in line 101 and 103 also do a `.toString()` for each rule value.

- removing the check for `subtargetEl.checked`

In the line 100 this check skipped all rule tests if a checkbox is checked and if there is a `data-dependent` with a `false` value: it did not do the check.

_Note: I am not aware of other side effects with this change._

```javascript
(subtargetEl.checked && !subtargetEl.matches('[type="radio"]'))
```



Draft:

![Bildschirmfoto 2024-08-10 um 23 18 11](https://github.com/user-attachments/assets/61e62742-34ee-4b3c-8d5a-acf04beeec3f)
![Bildschirmfoto 2024-08-10 um 23 17 56](https://github.com/user-attachments/assets/79a84603-a9c5-4a80-8ccc-6fac0f726129)

muescha commented:
I tried with radio boxes but they are not that nice because I can not write in one line:
`- time pattern: (x) custom ( ) templates`

```js

      time_remaining_format_use_custom_radio_1: {
         _tagName: 'input',
         label: 'Use custom template',
         type: 'radio',
         name: `time_remaining_format_use_custom_radio`,
         checked: true,
         value: `true`,
      },
      time_remaining_format_use_custom_radio_2: {
         _tagName: 'input',
         label: 'Use pattern templates',
         type: 'radio',
         name: `time_remaining_format_use_custom_radio`,
         value: `false`,
      },
```
      
![Bildschirmfoto 2024-08-10 um 23 56 57](https://github.com/user-attachments/assets/c7b7357f-3fda-47be-bd9b-258552896456)
![Bildschirmfoto 2024-08-10 um 23 56 49](https://github.com/user-attachments/assets/57dfe055-dbc0-4887-b4a5-3a985c5497ff)

_note: I don't found code snippets about radio so I added some info to the wiki (https://github.com/raingart/Nova-YouTube-extension/wiki/plugin#radio)_

raingart commented:
1. originally a function `attrDependencies` ([code](https://github.com/raingart/Nova-YouTube-extension/blob/master/js/libs/options.js#L60-L129)) had only one function: hide all options when the plugin is deactivated.
2. Then there was a need to display options that depend on other (parent) options.
3. Adding inversion by specifying the `!` sign at the beginning [code](https://github.com/raingart/Nova-YouTube-extension/blob/master/plugins/control-panel/time-jump.js#L570)
4. And the last iteration is adding a similar function and turning off conflicting functions [code](https://github.com/raingart/Nova-YouTube-extension/blob/master/js/optionsBilder.js#L472-L512)

my short-sightedness led to the fact that instead of some universal method there was a modification of the `attrDependencies` that was initially not designed for this module.
- for example, to work with the `select` tag you need to specify an array
- for `checkbox` you need to specify Boolean parameters due to conversion when saving (I don’t remember why I did this) [code](https://github.com/raingart/Nova-YouTube-extension/blob/master/js/libs/options.js#L143-L147). At the same time, the inversion of checkbox (`!`) For Boolean type became naturally unavailable
- while this inversion works for string types. For example, the `select` tag does not convert [code](https://github.com/raingart/Nova-YouTube-extension/blob/master/plugins/player/volume.js#L404)

that is, such use cases are available
- checkbox - `'data-dependent': { 'parent_checkbox_name': Boolean }`,
- select - `'data-dependent': { 'parent_select_name': ['String', 'String-Number'], '!String' }`,
- input (type:number, text) - `'data-dependent': { 'parent_input_name': "0" },` and `'data-dependent': { 'parent_input_name': "!0" },`

it is also possible to "overlap" options. Because they are processed sequentially. You can make a mistake where two or more options may not work correctly.

Here is the entire list of data formats that I know and use that are saved:
![Screenshot from 2024-08-11 11-45-05](https://github.com/user-attachments/assets/f422bb53-1ba0-462c-98a5-512b67250db3)

The very idea of ​​implementing a more flexible module terrifies me. I don't have the slightest idea how to implement it. I compared it to the current implementation, which is very limited, but it already exists and somehow works.

for testing you can select the [speed](https://github.com/raingart/Nova-YouTube-extension/blob/master/plugins/player/speed.js) plugin covers almost all use cases

At the very beginning I tried to look for a ready-made solution, but I couldn’t find it. Perhaps if you put the instructions for LLM (GPT like) correctly, it will be able to generate a ready-made function. But I can't even describe the task. Since, as you previously indicated, the data storage system is not optimal and uses different data storage formats for similar types. 
For example, why are nets stored in String? The answer is simply because then the inversion (`!`) stops working. That is, one crippling implementation will ruin the subsequent ones

raingart commented:
So I found a wonderful error that confirms the fragility and non-obviousness of the current implementation

doesn't work completely. I probably added just the inverse of the empty string in desperation
https://github.com/raingart/Nova-YouTube-extension/blob/2437cb554a91c436ab3ad7bf91bd03729b5e5099/plugins/player/speed.js#L977

since the specified type is Boolean and not String
https://github.com/raingart/Nova-YouTube-extension/blob/2437cb554a91c436ab3ad7bf91bd03729b5e5099/plugins/player/speed.js#L744

need fix `{ label: 'none', value: 'false' },` in speed.js#L744

despite the fact that when saved from convects into a Boolean
`("NovaTube") "sync" : "undefined" => "{"rate_hotkey":false}"`

muescha commented:
Yes, the code is a bit fragile.

Initially, I attempted to use `'data-dependent': { 'time_remaining_format_use_custom': false }`.

During debugging, I discovered that setting this value to `true` always works. However, this is not due to the check in `|| ruleValues.some(i => currentValuesList.includes(i.toString()))`, but rather because of the shortcut `subtargetEl.checked`.

I also found that regardless of the checkbox state, the value is always `on`; there is no `off` value, nor is there an empty value. Even when unchecked, the value remains `on`.

To address this, I made the following change:

```diff
- return [subtargetEl.value];
+ return [(subtargetEl.type == 'checkbox') ? subtargetEl.checked.toString() : subtargetEl.value];
```

I added `toString()` to ensure all values are treated as strings.

With this change, there's no need to change the boolean values at 'data-dependent' to string, since all values are now strings (thanks to `subtargetEl.checked.toString()` and `currentValuesList.includes(i.toString())`).

Additionally, I removed this check:

```diff
- (subtargetEl.checked && !subtargetEl.matches('[type="radio"]')) // skip radio (which is always checked. Unlike a checkbox)
+ false //(subtargetEl.checked && !subtargetEl.matches('[type="radio"]')) // skip radio (which is always checked. Unlike a checkbox)
```

The reason for this removal is that, otherwise, the code never reaches the negation check. I tested using `!on` and `!true`, but this line prevented the negation check when the checkbox was checked. I didn't account for any side effects with this change, so it may need to be reviewed and fine-tuned further.

muescha commented:
BTW:

I played a little with the radio and added a fieldset: 

![Bildschirmfoto 2024-08-11 um 19 26 05](https://github.com/user-attachments/assets/c8e31855-f229-4f61-baa5-c3a0e5e7c55b)

![Bildschirmfoto 2024-08-11 um 19 24 57](https://github.com/user-attachments/assets/36921eaf-3bdd-44cd-9ea2-82ff800141ad)

this can be done if we have a `_tagname: 'fieldset'` element where the `legend: 'abc'` creates a `legend` html element, and we allow a field with `children` where all the child values goes into. 

the style for the radio buttons would change to show the `( )` in front of the text.


here is the resulting html snippet for the screenshot:
```html
<!--

ul ul ul {
    list-style-type: none;
}

li>label:before {
    content: "↪";
    margin: 0 .5em;
    color: #343a45;
}

 -->

<li class="item">
    <div class="info" tooltip="Remaining time until the end of the video" flow="up">
        <label for="time-remaining">Remaining time</label>
        <a href="https://github.com/raingart/Nova-YouTube-extension/wiki/plugins#time-remaining" target="_blank"
           title="More info">?</a>

    </div>
    <div class="opt">
        <input type="checkbox" name="time-remaining" id="time-remaining">
    </div>
    <ul data-dependent="{&quot;time-remaining&quot;:[true]}">
        <li><!--<label for="time_remaining_format_use_custom_radio_fieldset">select a template</label>-->
            <fieldset legend="selects a template" name="time_remaining_format_use_custom_radio_fieldset"
                      id="time_remaining_format_use_custom_radio_fieldset" class="collapse">
                <legend>Select pattern template</legend>
                <ul>
                    <li class=""><input type="radio" name="time_remaining_format_use_custom_radio" checked="true"
                                        value="true" id="time_remaining_format_use_custom_radio_1" class=""><label
                            for="time_remaining_format_use_custom_radio_1" class="collapse">Custom</label></li>
                    <li style=""><input type="radio" name="time_remaining_format_use_custom_radio" value="false"
                                        id="time_remaining_format_use_custom_radio_2" class=""><label
                            for="time_remaining_format_use_custom_radio_2" class="collapse">Templates</label></li>
                </ul>

            </fieldset>
        </li>
        <li style="
display:none"><label for="time_remaining_format_use_custom_radio_1" style="
">Use custom template</label><input type="radio" name="time_remaining_format_use_custom_radio" checked="true"
                                    value="true" id="time_remaining_format_use_custom_radio_1" class=""></li>
        <li style="
display:none"><label for="time_remaining_format_use_custom_radio_2">Use pattern templates</label><input type="radio"
                                                                                                        name="time_remaining_format_use_custom_radio"
                                                                                                        value="false"
                                                                                                        id="time_remaining_format_use_custom_radio_2"
                                                                                                        class="collapse">
        </li>
        <li data-dependent="{&quot;time_remaining_format_use_custom_radio&quot;:true}"
            tooltip="[^] - correction current playback speed" class=""><label for="time_remaining_format">Custom time
            pattern</label><input type="text" list="time_remaining_format_help_list"
                                  placeholder="{done%}/{duration^} ({rate})" minlength="4" maxlength="100"
                                  required="true" name="time_remaining_format" id="time_remaining_format"
                                  class="collapse"></li>
        <li>
            <datalist name="time_remaining_format_help_list" id="time_remaining_format_help_list">
                <option value="{done}/{duration} ({done%})">0:10/0:44 (23%)</option>
                <option value="{done^}/{duration^} {rate}">0:05/0:25 1.75x</option>
                <option value="{left}/{duration} • {left%}">-0:34/0:44 • -77%</option>
                <option value="{left^}/{duration^} {rate}">-0:19/0:25 1.75x</option>
                <option value="{left}/{duration} • {left^}/{duration^} ({done%}) {rate}">-2:24/18:00 • -0:48/6:00 (87%)
                    3x
                </option>
                <option value="{left} {left^} {left%} • {done} {done^} {done%} • {duration} {duration^} • {rate}">-2:24
                    -0:48 -13% • 15:36 5:12 87% • 18:00 6:00 • 3x
                </option>
                <option value=" ">For a custom template, you can use these fields:</option>
                <option value="{rate}">1.75x</option>
                <option value="{left}">-0:34</option>
                <option value="{left^}">-0:19</option>
                <option value="{left%}">-77%</option>
                <option value="{done}">0:10</option>
                <option value="{done^}">0:05</option>
                <option value="{done%}">23%</option>
                <option value="{duration}">0:44</option>
                <option value="{duration^}">0:25</option>
            </datalist>
        </li>
        <li data-dependent="{&quot;time_remaining_format_use_custom_radio&quot;:false}"
            tooltip="[^] - correction current playback speed" class="hide"><label for="time_remaining_format_templates">Time
            pattern templates</label><select required="true" name="time_remaining_format_templates"
                                             id="time_remaining_format_templates" disabled="">
            <option value="{left^}">{left^}</option>
            <option value="{left^} ({done%})">{left^} ({done%})</option>
            <option value="{left^} ({left%})">{left^} ({left%})</option>
            <option value="{left}/{left^} ({done%})">{left}/{left^} ({done%})</option>
            <option value="{left}">{left}</option>
            <option value="{left%}">{left%}</option>
            <option value="{done%}">{done%}</option>
            <option value="{done^}/{left^}">{done^}/{left^}</option>
            <option value="{done^}/{duration^}">{done^}/{duration^}</option>
            <option value="{left^}/{duration^}">{left^}/{duration^}</option>
            <option value="{left^}/{duration^} ({done%})">{left^}/{duration^} ({done%})</option>
            <option value="{left}/{left^} ({done%}) {rate}">{left}/{left^} {rate}</option>
        </select></li>
        <li><label for="time_remaining_position">Render section</label><select name="time_remaining_position"
                                                                               id="time_remaining_position">
            <option value="player">player</option>
            <option value="description">description</option>
        </select></li>
    </ul>
</li>
```

raingart commented:
I tried to implement `radio` elements at the initial stages
but I came to the conclusion that they are generally useless. This is a restriction analogue to `select`

it takes up more space and has no functional advantage

for example in Remaining time > Render section
or other places. select is easier to expand if necessary

muescha commented:
only advantage is that all options are better visible for the user. 
and yes - every select can be shown as radio or as select. 

If you don't like the radio thing, then I don't invest time to implement it :)
